File: frontend/src/services/storyService.ts
import api from './api';
import type { ApiResponse, StoryResponse, PageResponse, GenreResponse } from '../types';

export interface StoryFilters {
  keyword?: string;
  genreIds?: number[];
  status?: string;
  page?: number;
  size?: number;
  sort?: string;
}

export const storyService = {
  /**
   * Get stories with filtering and pagination
   */
  getStories: async (filters: StoryFilters = {}): Promise<ApiResponse<PageResponse<StoryResponse>>> => {
    const params = new URLSearchParams();

    if (filters.keyword) params.append('keyword', filters.keyword);
    if (filters.genreIds && filters.genreIds.length > 0) {
      params.append('genreIds', filters.genreIds.join(','));
    }
    if (filters.status) params.append('status', filters.status);
    params.append('page', String(filters.page ?? 0));
    params.append('size', String(filters.size ?? 12));
    params.append('sort', filters.sort ?? 'createdAt,desc');

    const response = await api.get<ApiResponse<PageResponse<StoryResponse>>>(`/stories?${params.toString()}`);
    return response.data;
  },

  /**
   * Get story detail by ID (includes chapter list)
   */
  getStoryById: async (id: number): Promise<ApiResponse<StoryResponse>> => {
    const response = await api.get<ApiResponse<StoryResponse>>(`/stories/${id}`);
    return response.data;
  },

  /**
   * Get all genres (for filter dropdown)
   */
  getGenres: async (): Promise<ApiResponse<GenreResponse[]>> => {
    const response = await api.get<ApiResponse<GenreResponse[]>>('/genres');
    return response.data;
  },

  /**
   * Admin: Create story
   */
  createStory: async (data: import('../types').StoryRequest): Promise<ApiResponse<StoryResponse>> => {
    const response = await api.post<ApiResponse<StoryResponse>>('/admin/stories', data);
    return response.data;
  },

  /**
   * Admin: Update story
   */
  updateStory: async (id: number, data: import('../types').StoryRequest): Promise<ApiResponse<StoryResponse>> => {
    const response = await api.put<ApiResponse<StoryResponse>>(`/admin/stories/${id}`, data);
    return response.data;
  },

  /**
   * Admin: Delete story
   */
  deleteStory: async (id: number): Promise<ApiResponse<void>> => {
    const response = await api.delete<ApiResponse<void>>(`/admin/stories/${id}`);
    return response.data;
  },
};

File: backend/src/main/java/com/echonovel/service/StoryService.java
package com.echonovel.service;

import com.echonovel.dto.request.StoryRequest;
import com.echonovel.dto.response.StoryResponse;
import com.echonovel.enums.StoryStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

/**
 * Service interface for story operations.
 */
public interface StoryService {

    /**
     * Get all stories with filtering and pagination.
     */
    Page<StoryResponse> getStories(String keyword, List<Long> genreIds, StoryStatus status, Pageable pageable);

    /**
     * Get story by ID (includes chapters).
     */
    StoryResponse getStoryById(Long id);

    /**
     * Create a new story (Admin).
     */
    StoryResponse createStory(StoryRequest request);

    /**
     * Update a story (Admin).
     */
    StoryResponse updateStory(Long id, StoryRequest request);

    /**
     * Delete a story (Admin).
     */
    void deleteStory(Long id);

    /**
     * Upload and update story cover image via Cloudinary.
     */
    StoryResponse updateCoverImage(Long storyId, MultipartFile file);

    /**
     * Get recommended stories with same genres (exclude current story).
     */
    List<StoryResponse> getRecommendations(Long storyId);
}
